scipy/stats/_hypotests.py

```diff
@@ -9,11 +9,101 @@
 from ._continuous_distns import chi2, norm
 from scipy.special import gamma, kv, gammaln
 from . import _wilcoxon_data
+import scipy.stats.stats
+from functools import wraps
+
+
+def _remove_nans(samples, paired):
+    "Remove nans from pair or unpaired samples"
+    # consider usnot copying arrays that don't contain nans
+    if not paired:
+        return [sample[~np.isnan(sample)] for sample in samples]
+        # return [sample[~np.isnan(sample)] if contains_nan else sample
+        #         for sample, contains_nan in zip(samples, contains_nans)]
+    # else
+    nans = np.isnan(samples[0])
+    for sample in samples[1:]:
+        nans = nans | np.isnan(sample)
+    not_nans = ~nans
+    return [sample[not_nans] for sample in samples]
+
+
+def _vectorize_hypotest_factory(result_creator, default_axis=0,
```

Comment:
I am not sure about having this decorator doing 2 things: nan and vectorization. On the other hand, having 2 separate decorator would mean we have to be careful about the order they are used on the function.

Reply:
I thought about splitting them (see [here](https://github.com/scipy/scipy/pull/13312#issuecomment-803383683)). It's been a while since I've looked at this, but there might have been some optimizations I got here because they were combined. I

Reply:
No no I am not asking to split. I just wanted to make sure you considered it and got some input here.

Because in the case both are mixed, we cannot just add the nan policy without the vectorized part (well we can, but the vectorization part is not necessary then and still requires some overhead in passing some arguments with the decorator). But if there would not be a use case for that, no problem for keeping them together. In the end, as you said, we should probably implement this on each functions individually for speed reasons.

Reply:
If the function is vectorized but needs a `nan_policy`... well it really also needs to be vectorized _like this_ in order to handle `nan_policy='omit'` and vectorization at the same time. We could avoid the slowdown by accepting an argument (e.g. `vectorized`) in the decorator function that just uses the original function as-is when there are no NaNs in the input, maybe.

If the function already has `nan_policy` and just needs to be vectorized, well... no harm in applying this decorator over it, actually. The `nan_policy` part is probably _faster_ than that of the original function, especially those that rely on masked arrays, and this will guarantee consistent/correct behavior across all functions that use the decorator.